### DISCUSSION TITLE
How does PocketBase handle BigInt (64-bit) integers?

### DISCUSSION THREAD

**Discussion by voxxit** [pocketbase/pocketbase#341] [Q&A]
I am seeing things in the `JSON` type such as:

<img width="278" alt="Screen Shot 2022-08-21 at 10 15 58 PM" src="https://user-images.githubusercontent.com/12201/185844323-52c4a04c-9c8f-4ca6-b395-db3c6fb17560.png">
 
Which does not seem correct...

**Comment by ganigeorgiev** [comment]
PocketBase uses the standard go library for serializing/deserializing JSON, which means that by default any number in a JSON string will be decoded as double precision floats (aka. float64).

For consistency, the same rule applies when type `number` field is used - it will be normalized/casted to float64:
https://github.com/pocketbase/pocketbase/blob/master/models/schema/schema_field.go#L227-L228

It is possible to provide a custom JSON decoder for the `json` type field and serialize the result value as string but I'm not sure if that will be expected from most users (in JavaScript numbers are also floats between -+2^53-1; there is a special bigint type but I'm not sure how that works with the browsers `JSON.stringify/parse` functions).

_I'll move the issue to Q&A discussion._

**Reply by matancev** [reply]
Actually, pocketbase's `number` fields use `real` type in sqlite (8-byte floating point value) that leads to:
1) Cannot work with integers seamlessly. For example, if I have custom model with `int64` field, I can save it successfully, but reading existing records causes `converting driver.Value type float64 to a int64: invalid syntax`. This behavior is not obvious.
2) Silent loss of precision when trying to insert big integers. For example I cannot store facebook object ids safely (they are 64-bit integers).

Do I understand correctly, that integers are not supported because javascript's ints are actually floats and Pocketbase targets web projects only?

**Reply by ganigeorgiev** [reply]
@matancev I'm not sure that I understand your question. As mentioned, PocketBase uses `float64`, (not `int64`) for storing numbers (and yes, the used sqlite column type is `REAL`).

> but reading existing records causes converting driver.Value type float64 to a int64: invalid syntax

Could you provide an example when this is throwing the error? By default all record field values should be safely casted to float64 when loaded in `models.Record`
https://github.com/pocketbase/pocketbase/blob/dba66d4da14455a00667c1115d4384dbdd76ac88/models/schema/schema_field.go#L255
If you want to get the value as int, you could use `record.GetInt("myField")`

> Silent loss of precision when trying to insert big integers.

If you have to store bigint values my recommendation would be to use the `text` type field. Even if we send int64 in the serialized JSON, it most likely won't be deserialized as expected in most languages (including JS).

**Reply by ganigeorgiev** [reply]
On second read, I understand that you are using a custom model and I guess in this case it is possible the above error to occur if your struct has int64 field. The workaround would be to create your own Go type that can safely cast the float64 to int64 (but there could be some precision lost for 53+ bit integers).
Depending on the users demand, eventually we can add a int64 option to the `number` field type (or a new `int` field type), but I think this will be confusing to most users and as mentioned previously the json deserialization may not work as expected everywhere.
